bootstrap/deploy_job.yaml

```diff
@@ -0,0 +1,31 @@
+# Deploy kubeflow to namespace "kubeflow"
+
+# TODO(inc0): Create rbac role bindings
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: kubeflow
+
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: kubeflow-deploy
+  namespace: kubeflow
+spec:
+  template:
+    spec:
+      containers:
+      - name: kubeflow-deploy
+        image: gcr.io/kubeflow-images-staging/bootstraper:latest
+        env:
+        - name: PVC_MOUNT_PATH
```

Comment:
Isn't this just a layer of indirection? e.g. you're taking the parameters exposed by ksonnet and just turning them into environment variables? Why not just tell users to run ks param set if they want to set it.

The point of the bootstraper is to solve the problem of Kubeflow exposes X parameters how do I set those parameters optimally for my deployment?

Forcing users to manually select those parameters (whether as environment variables or ks parameters) doesn't solve that problem. All this is doing is telling users that instead of doing `ks param set` they need to edit key-value paris in a YAML file. 

The bootstrapper is trying to do something fundamentally different to create a different experience. In particular, the boot strapper figures out based on the users setup how to set different values.

The current bootstrapper has a pretty simplistic example. The bootstrapper will check if the user has PVC and depending on the result set the value of jupyterNotebookPVCMount accordingly.

So setting jupyterNotebookPVCMount should be unnecessary except in the case the user wants to override it for some reason.

Reply:
There are few options how to approach that. We can create set of ENV like here, easiest. Better way would be to create configmap + shim layer to translate optiosn from configmap to params. Best way would be to have ksonnet app definition with all the configurations available (can be created by bootstrapper) and then consume it in job, so user can do sth like that:

1. bootstrapper - create template ksonnet app in app.yaml
2. edit app.yaml to your taste
3. kubectl create -f app.yaml
4. kubectl create -f https://raw.githubusercontent.com/kubeflow/kubeflow/master/bootstrap/deploy_job.yaml

And, in case you just want to roll with defaults, only step 4 is needed.

Reply:
Why do we need another shim layer?

What is the long term plan? Would there be a 1:1 mapping from environment variables to ksonnet parameters? 

Is the fact that the environment variable name is different from the ksonnet parameter intentional? Or is it just an oversight?

Reply:
oversight/trying to make it look like ENV variable rather than camel-cased one.

envs is just one way to pass params to deployment and I'm by no means saying it's best one. I think best one would be bootstraper-driven CRD or configmap.

End goal I think would be something like this:
run boostraper locally or in container (since it's golang app it can totally be just wget->run).
bootstraper will create CRD or configmap or any other k8s resource definition (ksonnet/app?) that will contain all the params it figures out are best - I'd say it should output editable file or something to allow operator modify them prior to running deployment.
operator runs deploy_job which will look for bootstraper-driven configmap and deploy accordingly.

Now if we also add safe defaults to deploy_job, bootstraper steps are optional for super-quickstart